Sources/GRPC/ConnectionPool/HTTP2ConnectionState.swift

```diff
@@ -0,0 +1,259 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Logging
+import NIO
+import NIOHTTP2
+
+/// This struct models the state of an HTTP/2 connection and provides the means to indirectly track
+/// active and available HTTP/2 streams on that connection.
+///
+/// The state -- once ready -- holds a multiplexer which it yields when an available 'token' is
+/// borrowed. One token corresponds to the creation of one HTTP/2 stream. The caller is responsible
+/// for later returning the token.
+internal struct HTTP2ConnectionState {
+  /// An identifier for this pooled connection.
+  internal let id: ObjectIdentifier
+
+  /// Indicates whether the pooled connection is idle.
+  internal var isIdle: Bool {
+    return self.state.isIdle
+  }
+
+  /// The number of tokens currently available for this connection.
+  internal var availableTokens: Int {
+    switch self.state {
+    case let .ready(ready):
+      return ready.availableTokens
+    case .idle, .connectingOrBackingOff:
+      return 0
+    }
+  }
+
+  /// The number of tokens currently borrowed from this connection.
+  internal var borrowedTokens: Int {
+    switch self.state {
+    case let .ready(ready):
+      return ready.borrowedTokens
+    case .idle, .connectingOrBackingOff:
+      return 0
+    }
+  }
+
+  /// The state of the pooled connection.
+  private var state: State
+
+  private enum State {
+    /// No connection has been asked for, there are no tokens available.
+    case idle
+
+    /// A connection attempt is underway or we may be waiting to attempt to connect again.
+    case connectingOrBackingOff
+
+    /// We have an active connection which may have tokens borrowed.
+    case ready(ReadyState)
+
+    /// Whether the state is `idle`.
+    var isIdle: Bool {
+      switch self {
+      case .idle:
+        return true
+      case .connectingOrBackingOff, .ready:
+        return false
+      }
+    }
+  }
+
+  private struct ReadyState {
+    internal var multiplexer: HTTP2StreamMultiplexer
+    internal var borrowedTokens: Int
+    internal var tokenLimit: Int
+
+    internal init(multiplexer: HTTP2StreamMultiplexer) {
+      self.multiplexer = multiplexer
+      self.borrowedTokens = 0
+      // 100 is a common value for HTTP/2 SETTINGS_MAX_CONCURRENT_STREAMS so we assume this value
+      // until we know better.
+      self.tokenLimit = 100
+    }
+
+    internal var availableTokens: Int {
+      return self.tokenLimit - self.borrowedTokens
+    }
+
+    internal mutating func borrowTokens(_ count: Int) -> (HTTP2StreamMultiplexer, Int) {
+      self.borrowedTokens += count
+      assert(self.borrowedTokens <= self.tokenLimit)
+      return (self.multiplexer, self.borrowedTokens)
+    }
+
+    internal mutating func returnToken(multiplexerID: ObjectIdentifier) {
+      if multiplexerID == ObjectIdentifier(self.multiplexer) {
```

Comment:
I revisited this; I actually don't think this can happen anymore

Reply:
You're not sounding super confident - turn into an assert for a while to raise confidence?

Reply:
Ah sorry, missed this comment @PeterAdams-A -- I'm confident it can't happen the way I initially thought it could so I removed the check. I'll add an assertion that we don't over-return tokens though which equates to the same thing.